pulp_docker/app/views.py

```diff
@@ -14,3 +57,235 @@ def get(self, request):
+        """
+        Return a response to the "GET" action.
+        """
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        try:
+            ca = ContentArtifact.objects.get(
+                content__in=distribution.publication.repository_version.content,
+                relative_path=tag_name)
+            content = ca.content.cast()
+            if content.manifest:
+                headers = {'Content-Type': content.manifest.media_type}
+            else:
+                headers = {'Content-Type': content.manifest_list.media_type}
+
+        except MultipleObjectsReturned:
+            log.debug(
+                _('Multiple (pass-through) matches for {b}/{p}'),
+                {
+                    'b': distribution.base_path,
+                    'p': tag_name,
+                }
+            )
+
+        except ObjectDoesNotExist:
+            pass
+        else:
+            artifact = ca.artifact
+            if artifact:
+                return self._dispatch(artifact.file.name, headers)
+            else:
+                raise ArtifactNotFound(path)
+
+        raise PathNotResolved(path)
+
+
+class BlobManifestView(views.APIView, ServeContentMixin):
+    """
+    APIView for Docker Registry v2 blob and manifest APIs
+    """
+
+    authentication_classes = []
+    permission_classes = []
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def get(self, request, path, digest):
+        """
+        Return a response to the "GET" action.
+        """
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        try:
+            ca = ContentArtifact.objects.get(
+                content__in=distribution.publication.repository_version.content,
+                relative_path=digest)
+            headers = {'Content-Type': ca.content.cast().media_type}
+        except MultipleObjectsReturned:
+            log.debug(
+                _('Multiple (pass-through) matches for {b}/{p}'),
+                {
+                    'b': distribution.base_path,
+                    'p': hash,
+                }
+            )
+        except ObjectDoesNotExist:
+            pass
+        else:
+            artifact = ca.artifact
+            if artifact:
+                return self._dispatch(artifact.file.name, headers)
+            else:
+                raise ArtifactNotFound(path)
+
+        raise PathNotResolved(path)
+
+
+class TagsListView(views.APIView):
+    """
+    APIView for Docker Registry v2 tags/list API
+    """
+
+    authentication_classes = []
+    permission_classes = []
+    content_negotiation_class = IgnoreClientContentNegotiation
+
+    def get(self, request, path):
+        """
+        Return JSON of all tags in this repo
+        """
+
+        distribution = get_object_or_404(DockerDistribution, base_path=path)
+        tags = {'name': path, 'tags': []}
+        for c in distribution.publication.repository_version.content:
```

Comment:
I see what you mean about the benefit of the single-table content here. That will be much improved. Even if we don't end up doing that, it seems like we could do better. Maybe:

    Tag.objects.filter(content__in=distribution.pub.rv.content)